src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,358 @@
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-data-model-provider/EmberDataBuffer.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage-null-handling.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLVTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <protocols/interaction_model/Constants.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <limits>
+
+namespace chip {
+namespace app {
+namespace Ember {
+
+namespace {
+
+/// Maximum length of a string, inclusive
+///
+/// For nullable strings, the maximum length marker is used as a "NULL" marker
+constexpr uint32_t MaxLength(EmberAttributeBuffer::PascalString s, bool nullable)
+{
+    if (s == EmberAttributeBuffer::PascalString::kShort)
+    {
+        return nullable ? std::numeric_limits<uint8_t>::max() - 1 : std::numeric_limits<uint8_t>::max();
+    }
+    // EmberAttributeBuffer::PascalString::kLong:
+    return nullable ? std::numeric_limits<uint16_t>::max() - 1 : std::numeric_limits<uint16_t>::max();
+}
+
+struct UnsignedDecodeInfo
+{
+    unsigned byteCount;
+    uint64_t maxValue;
+};
+
+constexpr UnsignedDecodeInfo GetUnsignedDecodeInfo(EmberAfAttributeType type)
+{
+
+    switch (type)
+    {
+    case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return UnsignedDecodeInfo{ 1, 0xFF };
+    case ZCL_INT16U_ATTRIBUTE_TYPE: // Unsigned 16-bit integer
+        return UnsignedDecodeInfo{ 2, 0xFFFF };
+    case ZCL_INT24U_ATTRIBUTE_TYPE: // Unsigned 24-bit integer
+        return UnsignedDecodeInfo{ 3, 0xFFFFFF };
+    case ZCL_INT32U_ATTRIBUTE_TYPE: // Unsigned 32-bit integer
+        return UnsignedDecodeInfo{ 4, 0xFFFFFFFF };
+    case ZCL_INT40U_ATTRIBUTE_TYPE: // Unsigned 40-bit integer
+        return UnsignedDecodeInfo{ 5, 0xFFFFFFFFFF };
+    case ZCL_INT48U_ATTRIBUTE_TYPE: // Unsigned 48-bit integer
+        return UnsignedDecodeInfo{ 6, 0xFFFFFFFFFFFF };
+    case ZCL_INT56U_ATTRIBUTE_TYPE: // Unsigned 56-bit integer
+        return UnsignedDecodeInfo{ 7, 0xFFFFFFFFFFFFFF };
+    case ZCL_INT64U_ATTRIBUTE_TYPE: // Unsigned 64-bit integer
+        return UnsignedDecodeInfo{ 8, 0xFFFFFFFFFFFFFFFF };
+    }
+    chipDie();
+}
+
+struct SignedDecodeInfo
+{
+    unsigned byteCount;
+    int64_t minValue;
+    int64_t maxValue;
+
+    template <typename T>
+    static constexpr SignedDecodeInfo From()
+    {
+        return SignedDecodeInfo{
+            sizeof(T),
+            std::numeric_limits<T>::min(),
+            std::numeric_limits<T>::max(),
+        };
+    }
+
+    static constexpr SignedDecodeInfo OfSize(unsigned n)
+    {
+        return SignedDecodeInfo{
+            n,
+            -static_cast<int64_t>(1LL << (8 * n - 1)),
+            (static_cast<int64_t>((1LL << (8 * n - 1))) - 1),
```

Comment:
- Shifting into the sign bit for positive integer is UB.
- Cannot negate 1LL << (8 * 8 - 1) --> that shifts into the sign bit.

Suggest making a special case for 8.

Reply:
This is to handle odd sized integers only - so 7 is the max I need.

Reply:
I will update it to a template and a static_assert to avoid this wrong usage.